tests/cli.rs

```diff
@@ -0,0 +1,225 @@
+            .arg(&path)
+            .arg("--addr")
+            .arg("127.0.0.1:0")
+            .spawn()?
+    };
+
+    // wrap in `ProvideProcess` to ensure the spawned process is killed on drop
+    let mut provider = ProvideProcess { child: provider };
+    let stderr = provider.child.stderr.take().unwrap();
+    let stderr = BufReader::new(stderr);
+
+    // test provide output & get all in one ticket from stderr
+    let all_in_one = match_provide_stderr(stderr, use_stdin)?;
+
+    // create a `get-ticket` cmd & optionally provide out path
+    let mut cmd = Command::new(iroh);
+    cmd.arg("get-ticket").arg(all_in_one);
+    let cmd = if let Some(ref out) = out {
+        cmd.arg("--out").arg(out)
+    } else {
+        &mut cmd
+    };
+
+    // test get stderr output
+    let get_output = cmd.output()?;
+    assert!(get_output.status.success());
+    match_get_stderr(get_output.stderr)?;
+
+    // test output
+    match out {
+        None => {
+            let expect_content = std::fs::read(path)?;
+            assert_eq!(expect_content, get_output.stdout);
+            Ok(())
+        }
+        Some(out) => compare_files(path, out),
+    }
+}
+
+// Wrapping the [`Child`] process here allows us to impl the `Drop` trait ensuring the provide
+// process is killed when it goes out of scope.
+struct ProvideProcess {
+    child: Child,
+}
+
+impl Drop for ProvideProcess {
+    fn drop(&mut self) {
+        self.child.kill().ok();
+        self.child.try_wait().ok();
+    }
+}
+
+fn compare_files(expect_path: impl AsRef<Path>, got_dir_path: impl AsRef<Path>) -> Result<()> {
+    let expect_path = expect_path.as_ref();
+    let got_dir_path = got_dir_path.as_ref();
+    if expect_path.is_dir() {
+        let paths = std::fs::read_dir(expect_path)?;
+        for entry in paths {
+            let entry = entry?;
+            compare_files(entry.path(), got_dir_path)?;
+        }
+    } else {
+        let file_name = expect_path.file_name().unwrap();
+        let expect = std::fs::read(expect_path)?;
+        let got = std::fs::read(got_dir_path.join(file_name))?;
+        assert_eq!(expect, got);
+    }
+
+    Ok(())
+}
+
+// Looks for regex matches on stderr output for the getter.
+//
+// Errors on the first regex mis-match or if the stderr output has fewer lines than expected
+fn match_get_stderr(stderr: Vec<u8>) -> Result<()> {
+    let stderr = std::str::from_utf8(&stderr[..])?;
+    let mut lines = stderr.lines();
+    let res = vec![
+        r"Fetching: [\da-z]{59}",
+        r"\[1/3\] Connecting ...",
+        r"\[2/3\] Requesting ...",
+        r"\[3/3\] Downloading collection...",
+        r"\d* file\(s\) with total transfer size [\d.]* ?[KMGT]?i?B",
+        r"Done in \d* seconds?",
+    ];
+
+    for re in res {
+        matches(
+            lines
+                .next()
+                .context("Unexpected end of 'get' output")?
+                .trim(),
+            re,
+        )?;
+    }
+    Ok(())
+}
```

Comment:
maybe make a helper to iterate over the lines, so that writing assertions is a bit cleaner

```rust
assert_matches_lines!(
    stderr,
    r"Fetching: [\da-z]{59}",
    r"\[1/3\] Connecting ...",
    r"\[2/3\] Requesting ...",
    r"\[3/3\] Downloading collection...",
    r"\d* file\(s\) with total transfer size [\d.]* ?[KMGT]?i?B",
    r"Done in \d* seconds?",    
);
```

Reply:
Okay, so to "make it work" for both the `get` and `provide` use cases, I needed to be able to ensure we could check for a specific line a certain number of times, which is where this odd `r"expr"; 1` syntax comes from. 

It's more clear in `match_provide_stderr` below how it's intended to be used.

I'd love some input on:
1) if this is too weird
2) if there is a better way to do the macro. This is basically my first time writing one, so I'm sure there are things I missed.
3) if I should not use a macro and return it to the previous version.

Reply:
seems fine

Reply:
there is a need for multiple matches in `match_provider_output`

Reply:
I adjusted the macro accordingly! It definitely reads easier.

However, all of these matches (except for the all-in-one ticket) only require `is_match`, not `captures`. 

According to the `regex`, you should favor `is_match` over `captures` when you can, since `is_match` is less computationally heavy.

This is being used in testing, tho, so this is probably fine because it's probably better to lean on the side of legibility. But in case it matters, I thought I'd bring it up.

Reply:
Yeah, should be totally fine, given that these are not hot spots. Thanks for bringing it up though 👍